swent-epfl-groupe07/eduverse-android
# Add Profile Search & Management Features

## Overview
This PR implements a complete profile search and management system, including automatic profile creation with Google Auth integration and real-time username search functionality. The implementation includes comprehensive test coverage for both backend operations and UI components.

## Key Features
- **Profile Search System**
  - Real-time username-based search with debouncing
  - Material3 UI components with loading, empty, and error states
  - Display of user statistics (followers/following) in search results
  - Integration with dashboard navigation via topbar

- **Automated Profile Management**
  - Automatic profile creation on first visit using Google Auth data
  - Smart defaults using Google display name and photo
  - Username management with validation rules
  - Preservation of existing social features and publications

## Technical Implementation
### Repository Layer
- Added `searchProfiles` method with Firestore integration
- Implemented prefix-based username search
- Added pagination support via result limiting
- Optimized queries using Firestore's range operators

### ViewModel Layer
- Implemented `SearchProfileState` for search lifecycle management
- Added debounced search to prevent API overload
- Proper handling of concurrent searches with job cancellation
- Comprehensive error handling
- Added profile initialization logic with Google Auth integration

### UI Layer
- Created `SearchProfileScreen` with Material3 components
- Implemented real-time search results display
- Added proper state handling (loading/empty/error)
- Integrated user statistics display
- Added comprehensive UI test coverage with test tags

## Testing
### Backend Tests
- Updated `ProfileViewModelTest` for new profile creation flow
- Enhanced `ProfileRepositoryImplTest` with new profile operations
- Added test coverage for:
  - Profile creation and initialization
  - Username validation and uniqueness
  - Firestore operations
  - Social interactions
  - Search functionality
  - Publication handling

### UI Tests
- Added test tags to all search screen components
- Implemented comprehensive UI tests for:
  - Search functionality
  - Profile display
  - User interactions
  - State management

## Related Issues
- Closes #107 (Social features implementation)

## Testing Instructions
1. Run the complete test suite: `./gradlew test`
2. Test profile creation:
   - Log in with a new Google account
   - Verify automatic profile creation with Google data
3. Test search functionality:
   - Use the search bar in the dashboard
   - Verify real-time updates
   - Check empty state and error handling

## Screenshots
<table>
  <tr>
    <td><img src="https://github.com/user-attachments/assets/0b613f74-66ce-40ca-a118-a7891d9cd337" alt="Search Screen" /></td>
    <td><img src="https://github.com/user-attachments/assets/8375126d-020d-43c8-92b6-52c06d308dff" alt="Results Screen" /></td>
    <td><img src="https://github.com/user-attachments/assets/fd0b97f8-5a3d-44e4-945a-89f857630866" alt="Empty State Screen" /></td>
  </tr>
</table>
